src/Handler/TagHandler.php

```diff
@@ -112,9 +129,26 @@ public function addTags(array $tags)
      */
     public function invalidateTags(array $tags)
     {
-        $tagExpression = sprintf('(%s)(,.+)?$', implode('|', array_map('preg_quote', $this->escapeTags($tags))));
-        $headers = array($this->tagsHeader => $tagExpression);
-        $this->invalidator->invalidate($headers);
+        $escapedTags = array_map('preg_quote', $this->escapeTags($tags));
+
+        if (mb_strlen(implode('|', $escapedTags)) >= $this->headerLength) {
+            /*
+             * estimate the amount of tags to invalidate by dividing the max
+             * header length by the largest tag (minus 1 for the implode character)
+             */
+            $tagsize = max(array_map('mb_strlen', $escapedTags));
+            $elems = floor($this->headerLength / ($tagsize - 1)) ? : 1;
+
```

Comment:
we could do `} else { $elems = count($escapedTags); }`

and then have the foreach outside the if/else, to avoid the 3 redundant lines of code in the else below.
that should not cost much extra processing power, and avoids code duplication. there would be only one chunk so the foreach would only be executed once.